MapboxDirections/MBRouteOptions.swift

```diff
@@ -325,15 +325,23 @@ open class RouteOptions: NSObject, NSSecureCoding {
      - returns: A tuple containing an array of waypoints and an array of routes.
      */
     internal func response(_ json: JSONDictionary) -> ([Waypoint]?, [Route]?) {
-        let waypoints = (json["waypoints"] as? [JSONDictionary])?.map { waypoint -> Waypoint in
-            let location = waypoint["location"] as! [Double]
-            let coordinate = CLLocationCoordinate2D(geoJSON: location)
-            return Waypoint(coordinate: coordinate, name: waypoint["name"] as? String)
-        }
-        let routes = (json["routes"] as? [JSONDictionary])?.map {
-            Route(json: $0, waypoints: waypoints ?? self.waypoints, routeOptions: self)
+        if let waypoints = (json["waypoints"] as? [JSONDictionary]) {
+            let zippedWaypoints = zip(waypoints, self.waypoints).map { (json, waypoint) -> Waypoint in
+                let location = json["location"] as! [Double]
+                let coordinate = CLLocationCoordinate2D(geoJSON: location)
+                return Waypoint(coordinate: coordinate, name: waypoint.name ?? json["name"] as? String)
+            }
+            
+            let routes = (json["routes"] as? [JSONDictionary])?.map {
+                Route(json: $0, waypoints: zippedWaypoints, routeOptions: self)
+            }
+            return (zippedWaypoints, routes)
+        } else {
```

Comment:
You’re on the right track. I’d pull the `routes` definition out of the if/else block, like it was before.